src/core/opamSystem.ml

```diff
@@ -821,13 +821,344 @@ let lock_none = {
 let lock_isatleast flag lock =
   lock_max flag lock.kind = lock.kind
 
+(* read_lines processes the lines of a text file in a platform-agnostic manner
+   (which is to say that it preserves '\r' on all platforms). It has two modes
+   of operation - in the Lines mode, it returns a string list of the lines, in
+   CrLf mode, it returns true if every line ends "\r\n". *)
+type _ read_return = Lines : string list read_return
+                   | CrLf  : bool read_return
+
+let read_lines : type s . s read_return -> string -> s = fun mode file ->
+  let ch =
+    try open_in_bin file
+    with Sys_error _ -> raise (File_not_found file)
+  in
+  let read_lines f =
+    let rec read_lines cr acc =
+      match input_line ch with
+      | line ->
+          let length = String.length line in
+          let acc = f line acc in
+          read_lines (cr && length > 0 && line.[length - 1] = '\r') acc
+      | exception End_of_file ->
+          close_in ch;
+          let acc =
+            if cr then
+              List.rev_map (fun s -> String.sub s 0 (String.length s - 1)) acc
+            else
+              List.rev acc
+          in
+          (acc, cr)
+    in
+    read_lines true []
+  in
+  match mode with
+  | Lines ->
+      read_lines OpamStd.List.cons |> fst
+  | CrLf ->
+      read_lines (fun _ _ -> []) |> snd
+
+let translate_patch ~dir orig corrected =
+  (* CRLF detection with patching can be more complicated than that used here,
+     especially in the presence of files with mixed LF/CRLF endings. The
+     processing done here aims to allow patching to succeed on files which are
+     wholly encoded CRLF or LF against patches which may have been translated to
+     be the opposite.
+
+     The resulting patch will *always* have LF line endings for the patch
+     metadata (headers, chunk locations, etc.) but uses either CRLF or LF
+     depending on the target file. Endings in the patch are always preserved for
+     new files. The benefit of always using LF endings for the metadata is that
+     patch's "Stripping trailing CRs from patch" behaviour won't be triggered.
+
+     There are various patch formats, though only the Unified and Context
+     formats allow multiple files to be patched. I tired of trying to get
+     sufficient documented detail of Context diffs to be able to parse them
+     without resorting to reverse-engineering code. It is unusual to see them
+     these days, so for now opam just emits a warning if a Context diff file is
+     encountered and does no processing to it.
```

Comment:
Agreed, maybe just add a line to the doc where it explains the patch format (and that they must apply with `-p 1`) that they should be unified diffs.

Reply:
I updated the sentence in the docs